File: 93-restore-ip-addresses/93-restore-ip-addresses.cpp
class Solution {
public:
    vector<string> result;
    bool isValid(string& s){
        if(s.empty()){
            return false;
        }
        if(s.size() == 1){
            return true;
        }
        if(s[0] == '0'){
            return false;
        }
        int num = stoi(s);
        return num >= 0 && num <= 255;
    }
    void backtrack(int pos, int part, string ip, string& s){
        // base case
        if(pos == s.size()){
            if(part == 0){
                result.push_back(ip.substr(1));
            }
            return;
        }
        if(part == 0){
            return;
        }
        string tempIp = "";
        for(int i = pos; i < s.size(); i++){
            tempIp.push_back(s[i]);
            if(isValid(tempIp)){
                backtrack(i + 1, part - 1, ip + "." + tempIp, s);
            }
            else{
                break;
            }
        }
    }
    vector<string> restoreIpAddresses(string s) {
        if(s.size() > 12){
            return result;
        }
        backtrack(0, 4, "", s);
        return result;
    }
};

File: 93-restore-ip-addresses/93-restore-ip-addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = []
        
        
        def is_valid(s):
            if s is None:
                return False
            if len(s) == 1:
                return True
            if int(s[0]) == 0:
                return False
            return 0 <= int(s) <= 255
        
        
        def dfs(pos, part, path, s, res):
            if part == 0 and pos == len(s):
                res.append(path[1 :])
                return
            if part == 0:
                return
            for i in range(pos, len(s)):
                if is_valid(s[pos : i + 1]):
                    dfs(i + 1, part - 1, path + '.' + s[pos : i + 1], s, res)
                else:
                    break
                  
                
        dfs(0, 4, '', s, res)
        
        return res
    